modulus/utils/neighbor_list.py

```diff
@@ -0,0 +1,177 @@
+@wp.kernel
+def radius_search_count(
+    hashgrid: wp.uint64,
+    points: wp.array(dtype=wp.vec3),
+    queries: wp.array(dtype=wp.vec3),
+    result_count: wp.array(dtype=wp.int32),
+    radius: wp.float32,
+):
+    """
+    Warp kernel for counting the number of points within a specified radius
+    for each query point, using a hash grid for spatial queries.
+
+    Args:
+        hashgrid: An array representing the hash grid.
+        points: An array of points in space.
+        queries: An array of query points.
+        result_count: An array to store the count of neighboring points within the radius for each query point.
+        radius: The search radius around each query point.
+    """
+    tid = wp.tid()
+
+    # create grid query around point
+    qp = queries[tid]
+    query = wp.hash_grid_query(hashgrid, qp, radius)
+    index = int(0)
+    result_count_tid = int(0)
+
+    while wp.hash_grid_query_next(query, index):
+        neighbor = points[index]
+
+        # compute distance to neighbor point
+        dist = wp.length(qp - neighbor)
+        if dist <= radius:
+            result_count_tid += 1
+
+    result_count[tid] = result_count_tid
+
+
+@wp.kernel
+def radius_search_query(
+    hashgrid: wp.uint64,
+    points: wp.array(dtype=wp.vec3),
+    queries: wp.array(dtype=wp.vec3),
+    result_offset: wp.array(dtype=wp.int32),
+    result_point_idx: wp.array(dtype=wp.int32),
+    result_point_dist: wp.array(dtype=wp.float32),
+    radius: wp.float32,
+):
+    """
+    Warp kernel for performing radius search queries on a set of points,
+    storing the results of neighboring points within a specified radius.
+
+    Args:
+        hashgrid: An array representing the hash grid.
+        points: An array of points in space.
+        queries: An array of query points.
+        result_offset: An array to store the offset in the results array for each query point.
+        result_point_idx: An array to store the indices of neighboring points found within the radius for each query point.
+        result_point_dist: An array to store the distances to neighboring points within the radius for each query point.
+        radius: The search radius around each query point.
+    """
+    tid = wp.tid()
+
+    # create grid query around point
+    qp = queries[tid]
+    query = wp.hash_grid_query(hashgrid, qp, radius)
+    index = int(0)
+    result_count = int(0)
+    offset_tid = result_offset[tid]
+
+    while wp.hash_grid_query_next(query, index):
+        neighbor = points[index]
+
+        # compute distance to neighbor point
+        dist = wp.length(qp - neighbor)
+        if dist <= radius:
+            result_point_idx[offset_tid + result_count] = index
+            result_point_dist[offset_tid + result_count] = dist
+            result_count += 1
+
+
+def radius_search(
+    points: wp.array(dtype=wp.vec3),
+    queries: wp.array(dtype=wp.vec3),
+    radius: float,
+    grid_dim: Union[int, Tuple[int, int, int]] = (128, 128, 128),
```

Comment:
Since we now target Python 3.10+, you can do this:
```py
grid_dim: int | tuple[int, int, int] = (128, 128, 128)
```
and get rid of `Tuple` and `Union` imports.

Reply:
Switched to this type annotation